### PR TITLE
💸  USDT transaction payments

### PR DESCRIPTION
**Pull request by JuaniRios**
## What?
Allow extrinsic fee payment with AssetHub assets that have a price in the oracle

## Why?
Needed for OTM flow

## How?
- Add the pallet_asset_tx_payment which is tightly coupled to pallet_transaction_payment.
- It gets the PLMC fee required for an extrinsic, and converts it to the asset from pallet_assets with our own custom logic.
 
## Testing?
In the next PR

## Anything Else?
- There already were some existing generic structs that implemented this logic, but they assumed the tip and fee to go to the same place. Since we want the fee to go to the Blockchain Operation Treasury, and the tip to go to the block author, we had to reimplement most logic.
 
- The original struct was `FungiblesAdapter<CON, HC>`.
where CON was the converter and HC the crediting of the fee.
Our solution copies that, but also implements a third generic, such that the second pays the fee-tip, and the third the fee. For more info check `pallet_asset_tx_payment::payment::FungiblesAdapter`

### CODE REVIEW

**Comment by JuaniRios** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/403?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#403** <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/403?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#402** <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/402?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#401** <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/401?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#399** <a href="https://app.graphite.dev/github/pr/Polimec/polimec-node/399?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @JuaniRios and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: runtimes/polimec/src/lib.rs [hunk 1]
```diff
@@ -1152,6 +1158,34 @@ impl pallet_dispenser::Config for Runtime {
 	type WhitelistedPolicy = DispenserWhitelistedPolicy;
 }
 
+pub struct PLMCToFundingAssetBalance;
+impl ConversionToAssetBalance<Balance, AssetId, Balance> for PLMCToFundingAssetBalance {
+	type Error = InvalidTransaction;
+
+	fn to_asset_balance(plmc_balance: Balance, asset_id: AssetId) -> Result<Balance, Self::Error> {
```
Comment (lrazovic) [inline runtimes/polimec/src/lib.rs:1165]: Should we also check if the `asset_id` is sufficient?

File: runtimes/polimec/src/lib.rs [hunk 2]
```diff
@@ -1152,6 +1158,34 @@ impl pallet_dispenser::Config for Runtime {
...
+		let funding_asset_balance =
+			funding_asset_price.reciprocal().ok_or(InvalidTransaction::Payment)?.saturating_mul_int(usd_balance);
+		Ok(funding_asset_balance)
+	}
+}
+impl pallet_asset_tx_payment::Config for Runtime {
+	type Fungibles = ForeignAssets;
+	type OnChargeAssetTransaction = TxFeeFungiblesAdapter<
+		PLMCToFundingAssetBalance,
+		CreditFungiblesToAccount<AccountId, ForeignAssets, BlockchainOperationTreasury>,
```
Comment (lrazovic) [inline runtimes/polimec/src/lib.rs:1183]: _Note _ - I checked with the rest of the team: The `BlockchainOperationTreasury` is the right destination of the fees.

File: runtimes/polimec/src/lib.rs [hunk 3]
```diff
@@ -1152,6 +1158,34 @@ impl pallet_dispenser::Config for Runtime {
...
+			<PriceProviderOf<Runtime>>::get_decimals_aware_price(PLMC_FOREIGN_ID, USD_DECIMALS, PLMC_DECIMALS)
+				.ok_or(InvalidTransaction::Payment)?;
+		let funding_asset_decimals = <ForeignAssets as fungibles::metadata::Inspect<AccountId>>::decimals(asset_id);
+		let funding_asset_price =
+			<PriceProviderOf<Runtime>>::get_decimals_aware_price(asset_id, USD_DECIMALS, funding_asset_decimals)
+				.ok_or(InvalidTransaction::Payment)?;
+		let usd_balance = plmc_price.saturating_mul_int(plmc_balance);
+		let funding_asset_balance =
+			funding_asset_price.reciprocal().ok_or(InvalidTransaction::Payment)?.saturating_mul_int(usd_balance);
+		Ok(funding_asset_balance)
```
Comment (lrazovic) [inline runtimes/polimec/src/lib.rs:1176]: The logic should be good. The reciprocal always looks a bit scary, but it should be okay. Let's see how it behaves in the tests.

**Review by lrazovic — APPROVED** [review]

File: runtimes/shared-configuration/src/fee.rs [hunk 4]
```diff
@@ -148,3 +161,104 @@ where
...
+
+		// Calculate how much refund we should return.
+		let (final_fee, refund) = paid.split(converted_fee);
+		// Split the tip from the fee
+		let (final_fee_minus_tip, final_tip) = final_fee.split(converted_tip);
+
+		let _ = <Runtime::Fungibles as fungibles::Balanced<Runtime::AccountId>>::resolve(who, refund);
+
+		FeeCreditor::handle_credit(final_fee_minus_tip);
+		TipCreditor::handle_credit(final_tip);
```
Comment (lrazovic) [inline runtimes/shared-configuration/src/fee.rs:250]: This should be the main difference with the Parity's implementation, and it looks good, let's be sure to test the tip handling in the integration tests

**Review by lrazovic — APPROVED** [review]

**Review by vstam1 — APPROVED** [review]
lgtm as far as I understand the asset tx fee

**Pull request merged**